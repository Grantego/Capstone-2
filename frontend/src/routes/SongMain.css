html {
    box-sizing: border-box;
  }
  *,
  *:before,
  *:after {
    box-sizing: inherit;
  }
  

  html,
  body {
    height: 100%;
    margin: 0;
    line-height: 1.5;
    color: #121212;
  }

  #new-song-btn {
    margin-left: 1vw;
  }

  #song-header {
    text-decoration: none;
    color: black
  }

  #song-header:hover {  
    font-weight: bold; 
  }
  
  .sr-only {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    white-space: nowrap;
    border-width: 0;
  }
  
  #root {
    height: 100%;
    width: 100%;
  }
  
  #pagediv {
    display: flex;
    height: 100%;
    width: 100%;
  }
  #sidebar {
    width: 20vw;
    background-color: #f7f7f7;
    border-right: solid 1px #e3e3e3;
    display: flex;
    flex-direction: column;
  }
  
  #sidebar > * {
    padding-left: 2rem;
    padding-right: 2rem;
  }
  
  #sidebar h1 {
    font-size: 1rem;
    font-weight: 500;
    display: flex;
    align-items: center;
    margin: 0;
    padding: 1rem 2rem;
    border-top: 1px solid #e3e3e3;
    order: 1;
    line-height: 1;
  }

  #sidebar h5 {
    font-size: 2vw;
  }

  #new-song-btn {
    font-size: 1.5vw;
    width: 5vw;
    height: 3vw;
    display: flex;
    align-items: center;
  }
  
  #sidebar > div {
    display: flex;
    background-color: #e6a64c8a;
    align-items: center;
    justify-content: center;
    gap: 1vw;
    padding-top: 1rem;
    padding-bottom: 1rem;
    border-bottom: 1px solid #e3e3e3;
  }
  
  #sidebar nav {
    flex: 1;
    overflow: auto;
    padding-top: 1rem;
    padding-left: .5vw;
    padding-right: .5vw;
  }
  
  #sidebar nav a span {
    float: right;
    color: #eeb004;
  }
  #sidebar nav a.active span {
    color: inherit;
  }
  
  i {
    color: #818181;
  }
  #sidebar nav .active i {
    color: inherit;
  }
  
  #sidebar ul {
    padding: 0;
    margin: 0;
    list-style: none;
  }
  
  #sidebar li {
    margin: 0.25rem 0;
    font-size: 2vw;
  }
  
  #sidebar nav a {
    display: flex;
    align-items: center;
    justify-content: space-between;
    overflow: hidden;
    white-space: pre;
    padding: 0.5rem;
    border-radius: 8px;
    color: inherit;
    text-decoration: none;
    gap: 1rem;
  }
  
  #sidebar nav a:hover {
    background: #e3e3e3;
  }

  #detail {
    flex: 1;
    padding: 2rem 4rem;
    width: 100%;
  }


  #error-page {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 100%;
  }

  @media (max-width: 600px) {
    #sidebar li {
      margin: 0.25rem 0;
      font-size: 3vw;
    }

    #sidebar h5 {
      font-size: 3vw
    }
  }
  