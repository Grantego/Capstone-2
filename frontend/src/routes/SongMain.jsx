import React from "react"
import NavBar from "../components/NavBar"
import SpellerApi from "../../api"
import { Form, Outlet, NavLink } from "react-router-dom"
import { useLoaderData } from "react-router"
import { Button } from "reactstrap"
import "./SongMain.css"

export async function loader({params}) {
    if (localStorage.getItem('username') !== params.username) throw new Error("not authorized to view this page")
    let res = await SpellerApi.getUserSongs(params.username)
    return res
}

const SongMain = () => {
 const {songs} = useLoaderData()
    return (
        <>
        <NavBar/>
        <div id="pagediv">
            <div id="sidebar">
            <div>
                <h5><a id="song-header" href={`/${localStorage.getItem("username")}`}>My Songs</a></h5>
                <Button id="new-song-btn" href={`/${localStorage.getItem("username")}/new`}>New</Button>

            </div>
            <nav>
                <ul>
                    {songs.length ? songs.map((song) => (
                        <li key={song.id}>
                            <a href={`/${localStorage.getItem("username")}/${song.id}`}>{song.title}</a>
                        </li>
                    )) : <li>No songs found</li>}
                </ul>
            </nav>
            </div>
            <div id="detail"><Outlet/></div>
        </div>
      </>
  
    )
}

export default SongMain