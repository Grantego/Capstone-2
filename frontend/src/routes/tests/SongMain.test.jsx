import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import SongMain, {loader} from '../SongMain';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { describe, it, beforeEach, expect, vi } from 'vitest';
import SpellerApi from '../../../api';

vi.mock('../../../api',  () => {
    return {
        default: vi.fn().mockImplementation(() => {
            return {
                getUserSongs: vi.fn().mockResolvedValue({songs: [
                    {
                        id: 1,
                        title: "Test Song"
                    },
                    {
                        id: 2,
                        title: "Test Song 2"
                    },
                ]})                
            }
        })         
    }
});

describe('SongMain component', () => {
    let router

  beforeEach(() => {
    localStorage.clear();
    
    const mockLoader = async () => {
        const apiInstance = new SpellerApi();
        const res = await apiInstance.getUserSongs();
        console.log(res)
        return {songs: res.songs};
    }

    router = createMemoryRouter([
        {
            path: '/',
            element: <SongMain/>,
            loader: mockLoader
        },
    ], {
        initialEntries: ['/'],
    }); 


  });

  it("matches snapshot", () => {
    const {asFragment} = render(<RouterProvider router={router} />)
    expect(asFragment()).toMatchSnapshot();
  })

  it("renders without crashing", () => {
    render(<RouterProvider router={router} />)
  })

  it("renders all elements correctly", async () => {
    localStorage.setItem('username', 'testuser')
    localStorage.setItem('token', 'sampletoken')

    render(<RouterProvider router={router} />);

    await waitFor(() => {
        // Header elements
        expect(screen.getByText("My Songs")).toBeInTheDocument()
        expect(screen.getByText("New")).toBeInTheDocument()
        // Song List Elements
        expect(screen.getByText("Test Song")).toBeInTheDocument()
        expect(screen.getByText("Test Song 2")).toBeInTheDocument()
        
    })

  })
});