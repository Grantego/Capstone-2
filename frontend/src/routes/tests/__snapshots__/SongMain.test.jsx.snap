// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChordDetails component > matches snapshot 1`] = `<DocumentFragment />`;

exports[`SongMain component > matches snapshot 1`] = `<DocumentFragment />`;
